Add support for context fieldResolver

kamranayub opened the pull request:
Fixes https://github.com/apollographql/apollo-server/issues/716

Referenced: https://github.com/graphql/graphql-js/pull/865/files

Looks like we can use `context.fieldResolver` to access the custom field resolver if it exists. Tested with linked package and my tracing-enabled GQL requests now return data like I expect!

kamranayub commented:
Ah, hold up, I just need to verify something--with master or my change, I'm seeing an empty resolvers list in the tracing extension. Let me investigate that first to confirm my change didn't mess with anything.

kamranayub commented:
I don't know what's going on; if I use `npm link` even master breaks and resolvers never populate in tracing extension. If I don't use npm link and use published package, it works. This is outside of my change. I'd like to figure it out, though :(

kamranayub commented:
@martijnwalraven Do you know the best way to test this change? Do I need to link from the apollo-server package all the way down to my project?

martijnwalraven commented:
Thanks for working on this! Your changes look good to me. Testing unpublished dependent packages is a big pain, mostly because we can only have a single instance of `graphql-js` installed, and that doesn't work well with `npm link`.

kamranayub commented:
Oh ok, well, should we try it out then?

![image](https://user-images.githubusercontent.com/563819/34127127-9494f328-e401-11e7-93c6-cf4947ede548.png)

martijnwalraven commented:
Hmmm, sorry, I misspoke. I checked `graphql-js`, and it doesn't actually seem `fieldResolver` is exposed on `context`. The PR you referenced shows it as a property on `ExecutionContext`, but that is not the same as the context passed to resolvers.

kamranayub commented:
Oh! I wish I could test it locally then, so I can verify my changes :( Are there documented steps to get this working locally then?

martijnwalraven commented:
Unfortunately not. I've had some luck using Yarn workspaces together with Lerna, but I wouldn't be able to tell you the exact steps I followed.

kamranayub commented:
OK, I may keep exploring. From what I can tell, this might need to be a change in apollo-server-core or even better, maybe in here: https://github.com/graphql/graphql-js/blob/461392db1eb05b21dc8b048ac352ea2fb37bdbbd/src/type/definition.js#L1165

If the type config field map `field.resolve` took care of figuring out which resolve function to use, it would "just work" with the downstream Apollo packages. At least, based on my 20 minutes of digging.

martijnwalraven commented:
I'm afraid that wouldn't work, because `field.resolve` is the resolver you define as part of the schema, and a custom field resolver is something you can configure per execution.

kamranayub commented:
OK, so then the next place I looked was maybe some way to pass it through
via the apollo-server runQuery routine where it calls into
graphql-extensions maybe, as part of the building stack routine.

On Mon, Dec 18, 2017 at 9:51 PM Martijn Walraven <notifications@github.com>
wrote:

> I'm afraid that wouldn't work, because field.resolve is the resolver you
> define as part of the schema, and a custom field resolver is something you
> can configure per execution.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/graphql-extensions/pull/1#issuecomment-352632336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAiaa-GQ-4QbHK42I6_AI598PwNijKQhks5tBzK6gaJpZM4RF90D>
> .
>

martijnwalraven commented:
Yeah, I think that's probably the best approach for now.

kamranayub commented:
I'll try to take a look this week! It'll help us demo our GQL endpoint with
tracing to the people who pay my bills :)

On Mon, Dec 18, 2017 at 11:30 PM Martijn Walraven <notifications@github.com>
wrote:

> Yeah, I think that's probably the best approach for now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/graphql-extensions/pull/1#issuecomment-352644127>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAiaa16fSjeIczOinMv1_FpQAbyTCYKmks5tB0nqgaJpZM4RF90D>
> .
>

martijnwalraven commented:
Great! I might actually have a look at it myself today. I've taken a few days off, but I was planning on looking at some long overdue issues :)

Out of curiosity, what are you using the custom field resolver for?

kamranayub commented:
To by default look for snake_case property names. I know another (probably
better) approach is to handle that when getting responses from the API, so
GQL resolvers don't need to deal with it (since using `obj` I still deal
with snake_case props sometimes). But that's where we started and noticed
the bug! We might refactor it anyway but it'd be awesome to fix it too.

On Mon, Dec 18, 2017 at 11:37 PM Martijn Walraven <notifications@github.com>
wrote:

> Great! I might actually have a look at it myself today. I've taken a few
> days off, but I was planning on looking at some long overdue issues :)
>
> Out of curiosity, what are you using the custom field resolver for?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apollographql/graphql-extensions/pull/1#issuecomment-352645011>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAiaa2fXzRQ2PigNc1O9ZdpmUjP_vz17ks5tB0uXgaJpZM4RF90D>
> .
>

kamranayub commented:
I tried to timebox getting the dev pipeline set up and couldn't figure it out after 2 hours, so I went ahead and just removed our default resolver for now and replaced with the approach above. Too bad! If I would have figured it out, I would have documented what to do. 😢  I think maybe I'll close this for now.
